### DISCUSSION TITLE
Sketch of design requirements for a Provider use case

### DISCUSSION THREAD

**Discussion by jonahbeckford** [mbarbin/provider#37] [General]
*Transferred from https://github.com/mbarbin/provider/pull/14*

Context: I'm considering using Provider in [DkCoder](https://discuss.ocaml.org/t/dkcoder-0-2-scripting-in-ocaml/14560). The main simplifier is that I have structural information from [codept](https://github.com/Octachron/codept) about all the modules, module types and functors in a project. Another simplifier is that I can and do inject modules aliases on the top of each `.ml` script (like a PPX but with awareness of the whole project).

One of the last features to implement is conditional modules. For example, having `MyGame_DirectXBackend.ml` compile only on Windows and `MyGame_MetalBackend.ml` compile on macOS. And when it does compile, I'd want to pass the `MyGame_DirectXBackend` module a Windows DirectX implementation module and `MyGame_MetalBackend.ml` an Apple Metal framework module. Basically, do the wiring on behalf of the developer (similar to Spring wiring in Java programs).

I have an initial design based on matching functor parameter types (ex. `module Make (Mtl: REQUIRES_APPLE_METAL) = struct let device = Mtl.createSystemDefaultDevice() end`) with one or more corresponding module implementations (ex. `module Provides_APPLE_METAL () = struct ... end`).

When I get to finalizing the design next year, I might switch to use your Provider implementation. But one big difference is I can only take advantage of module/module type information (ex. the `Provider` and `Reader` modules were accessed in module `abc.ml`) rather than value/type information (ex. `[> reader ] Provider.package` is an argument to a function). I think I'd have to provide a high-level wrapper around `Provider.lookup` to perform lookups on a global Provider (or a Provider that I inject into each `.ml` script).

Anyway, that is my thoughts for now. I probably will ask more questions next year. But I'm open to suggestions before that.

**Comment by mbarbin** [comment]
Hi @jonahbeckford sorry it took me a while to get back to you! I know we chatted also in PRs and stuff. I wanted to give you some heads up, as I have edited the project README with a cautionary note where I express some doubts I have about this project future. I initially used it (vcs) and now use OCaml Objects, and have been pretty happy with that. So, I don't know where you are in your project, but perhaps `provider` isn't such a great move.. I don't know! That's how I feel now, and it's been the case for a few months. I added this link in the README https://gr-im.github.io/a/dependency-injection.html I wouldn't say this is an actual suggestion, more like evidence that other folks are thinking about using Objects for this kind of things. It's probably on your radar already! Cheers, good luck with DkCoder!
